src/bfs.cpp

```diff
@@ -3,56 +3,152 @@
-                //Instead this function will be replaced by taking the current loc
-                //the row/col/and box of the location of the guess and check if its possible
-                if(current.is_possible()) {
-                    next_iter.push_back(current);
+class BFS
+{
+public:
+    static Grid seq_solve(Grid g)
+    {
+        std::vector<Coordinate> coords = g.find_all_empty_cells();
+
+        std::deque<Grid> possible;
+        std::deque<Grid> next_iter;
+
+        possible.push_back(g);
+
+        // we will do the bfs solving by choosing one of the unsolved squres
+        // then we will find all possibilies for that square
+        // at any particular timestep, the coords is all the same
+        // pick and solve exactly one square in every full iteration of the while loop
+        while (!coords.empty())
+        {
+            Coordinate c = coords.back();
+            coords.pop_back();
+            next_iter.clear();
+
+            while (!possible.empty())
+            {
+                Grid private_possible = possible.front();
+                possible.pop_front();
+
+                for (dtype guess = 1; guess <= g.size(); guess++)
+                {
+                    // overriding value here (should be ok)
+                    private_possible[c.r][c.c] = guess;
+
+                    // this function is not efficient and should be replaced
+                    // Instead this function will be replaced by taking the private_possible loc
+                    // the row/col/and box of the location of the guess and check if its possible
+                    if (private_possible.is_possible())
+                    {
+                        next_iter.push_back(private_possible);
+                    }
+                }
+            }
+
+            possible.swap(next_iter);
+        }
+
+        if (!possible.empty())
+        {
+            return possible[0];
+        }
+
+        return Grid();
+    }
+
+    static Grid par_solve(Grid g)
+    {
+        std::vector<Coordinate> coords = g.find_all_empty_cells();
+
+        std::deque<Grid> possible;
+
+        possible.push_back(g);
+        dtype grid_size = g.size();
+
+        // we will do the bfs solving by choosing one of the unsolved squres
+        // then we will find all possibilies for that square
+        // at any particular timestep, the coords is all the same
+        // pick and solve exactly one square in every full iteration of the while loop
+        while (!coords.empty())
+        {
+            Coordinate c = coords.back();
+            coords.pop_back();
+            // next_iter.clear();
+
+#pragma omp parallel default(none) shared(c, possible, grid_size)
+            {
+                std::deque<Grid> next_iter;
+                int tid = omp_get_thread_num();
+                Grid private_possible;
+
+                while (true)
+                {
+#pragma omp critical
```

Comment:
This will probably be really slow. It would be nice if you could run some timing analysis on the code and see what is actually holding it up. Will need it for later anyway.

One suggestion for this is to remove this critical and the #pragma omp parallel on top
and change this into a parallel for

something like: 
#pragma parallel for 
for(size_t i = 0; i < possible.size(); i++) {
    Grid current = possible[i]; 
}

Then have a sync at the end. then Pragma single  (one thread) deletes the entire possible and then does a swap with next iter

Reply:
Brilliant idea! One concern I have is that if we have more possibles compared to number of threads, so we still need while(true), to have some thread come back and run the task again, right? But even so, it would be a faster approach than mine